File: reference/dotnet/src/YawLabs.Awsp/VerifyResult.cs
// AWSP -- A2A Webhook Security Profile reference implementation.

namespace YawLabs.Awsp;

/// <summary>
/// Result of <see cref="Awsp.Verify"/>.
/// <para>
/// On success, <see cref="Ok"/> is <c>true</c> and <see cref="Reason"/> is <c>null</c>.
/// On failure, <see cref="Ok"/> is <c>false</c> and <see cref="Reason"/> carries one of the
/// stable enum values defined by SPEC.md section 9:
/// </para>
/// <list type="bullet">
///   <item><description><c>malformed_header</c></description></item>
///   <item><description><c>unknown_algorithm</c></description></item>
///   <item><description><c>stale</c></description></item>
///   <item><description><c>future</c></description></item>
///   <item><description><c>replayed</c></description></item>
///   <item><description><c>unknown_kid</c></description></item>
///   <item><description><c>bad_hmac</c></description></item>
/// </list>
/// </summary>
/// <param name="Ok">True if verification succeeded.</param>
/// <param name="Reason">
/// Stable error enum on failure; <c>null</c> on success. Receivers MAY return this in
/// <c>{"error":"invalid_signature","reason":...}</c> JSON bodies but MUST NOT return any
/// free-form diagnostic.
/// </param>
public sealed record VerifyResult(bool Ok, string? Reason);

File: reference/java/src/main/java/com/yawlabs/awsp/VerifyResult.java
// AWSP -- A2A Webhook Security Profile reference implementation.
// Copyright 2026 Yaw Labs. Licensed under the Apache License, Version 2.0.

package com.yawlabs.awsp;

/**
 * Result of {@link Awsp#verify(VerifyParams)}.
 *
 * <p>{@code ok} is {@code true} on success and {@code reason} is {@code null}.
 * On failure {@code ok} is {@code false} and {@code reason} is one of the
 * stable spec enum strings:
 *
 * <ul>
 *   <li>{@code "malformed_header"}</li>
 *   <li>{@code "unknown_algorithm"}</li>
 *   <li>{@code "stale"}</li>
 *   <li>{@code "future"}</li>
 *   <li>{@code "replayed"}</li>
 *   <li>{@code "unknown_kid"}</li>
 *   <li>{@code "bad_hmac"}</li>
 * </ul>
 *
 * <p>Receivers SHOULD propagate {@code reason} to the 401 response body for
 * sender-side debuggability. Receivers in adversarial environments MAY
 * collapse all failures to {@code bad_hmac}. See SPEC.md section 9.
 */
public record VerifyResult(boolean ok, String reason) {

    /** Build a successful result. */
    public static VerifyResult success() {
        return new VerifyResult(true, null);
    }

    /** Build a failure result with the given stable reason code. */
    public static VerifyResult fail(String reason) {
        return new VerifyResult(false, reason);
    }
}
